tsdb/agent/storage.go

```diff
@@ -0,0 +1,678 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wal"
+	"go.uber.org/atomic"
+)
+
+var (
+	ErrUnsupported = errors.New("unsupported operation with WAL-only storage")
+)
+
+// Default values for options.
+var (
+	DefaultTruncateFrequency = 2 * time.Hour
+	DefaultMinWALTime        = 5 * time.Minute
+	DefaultMaxWALTime        = 4 * time.Hour
+)
+
+// Options of the WAL storage.
+type Options struct {
+	// Segments (wal files) max size.
+	// WALSegmentSize <= 0, segment size is default size.
+	// WALSegmentSize > 0, segment size is WALSegmentSize.
+	WALSegmentSize int
+
+	// WALCompression will turn on Snappy compression for records on the WAL.
+	WALCompression bool
+
+	// StripeSize is the size in entries of the series hash map. Reducing the size will save memory but impact performance.
+	StripeSize int
+
+	// SeriesLifecycleCallback specifies a list of callbacks that will be called during a lifecycle of a series.
+	// It is always a no-op in Prometheus and mainly meant for external users who import TSDB.
+	SeriesLifecycleCallback tsdb.SeriesLifecycleCallback
+
+	// TruncateFrequency determines how frequently to truncate data from the WAL.
+	TruncateFrequency time.Duration
+
+	// Shortest and longest amount of time data can exist in the WAL before being
+	// deleted.
+	MinWALTime, MaxWALTime time.Duration
```

Comment:
These should be `int64`. Look at my other comment about TSDB not knowing the unit of the timestamp.